src/server/frontend_wayland/foreign_toplevel_manager_v1.cpp

```diff
@@ -0,0 +1,696 @@
+    if (iter == surface_observers.end())
+    {
+        log_error("Can not remove ForeignSurfaceObserver: surface %p not in observers map");
+    }
+    else
+    {
+        iter->second->cease_and_desist();
+        surface_observers.erase(iter);
+    }
+}
+
+void mf::ForeignSceneObserver::surface_exists(std::shared_ptr<scene::Surface> const& surface)
+{
+    create_surface_observer(surface);
+}
+
+void mf::ForeignSceneObserver::end_observation()
+{
+    clear_surface_observers();
+}
+
+void mf::ForeignSceneObserver::create_surface_observer(std::shared_ptr<scene::Surface> const& surface)
+{
+    std::lock_guard<std::mutex> lock{mutex};
+    auto observer = std::make_shared<ForeignSurfaceObserver>(wayland_executor, manager, surface);
+    surface->add_observer(observer);
+    auto insert_result = surface_observers.insert(std::make_pair(surface, observer));
+    if (!insert_result.second)
+    {
+        log_error("Can not add ForeignSurfaceObserver: surface %p already in the observers map");
+        observer->cease_and_desist();
+    }
+}
+
+void mf::ForeignSceneObserver::clear_surface_observers()
+{
+    std::lock_guard<std::mutex> lock{mutex};
+    for (auto const& pair : surface_observers)
+    {
+        pair.second->cease_and_desist();
+        if (auto const surface = pair.first.lock())
+        {
+            surface->remove_observer(pair.second);
+        }
+    }
+    surface_observers.clear();
+}
+
+// ForeignSurfaceObserver
+
+mf::ForeignSurfaceObserver::ForeignSurfaceObserver(
+    std::shared_ptr<Executor> const& wayland_executor,
+    mw::Weak<ForeignToplevelManagerV1> manager,
+    std::shared_ptr<scene::Surface> const& surface)
+    : wayland_executor{wayland_executor},
+      manager{manager},
+      weak_surface{surface}
+{
+    std::lock_guard<std::mutex> lock{mutex};
+    create_or_close_toplevel_handle_as_needed(lock);
+}
+
+mf::ForeignSurfaceObserver::~ForeignSurfaceObserver()
+{
+    cease_and_desist();
+}
+
+void mf::ForeignSurfaceObserver::cease_and_desist()
+{
+    std::lock_guard<std::mutex> lock{mutex};
+    weak_surface.reset();
+    create_or_close_toplevel_handle_as_needed(lock);
+}
+
+void mf::ForeignSurfaceObserver::with_toplevel_handle(
+    std::lock_guard<std::mutex>&,
+    std::function<void(ForeignToplevelHandleV1&)>&& action)
+{
+    if (handle)
+    {
+        wayland_executor->spawn(
+            [handle = handle, action = std::move(action)]()
+            {
+                if (*handle)
+                {
+                    action(handle->value());
+                }
+            });
+    }
+}
+
+void mf::ForeignSurfaceObserver::create_or_close_toplevel_handle_as_needed(std::lock_guard<std::mutex>& lock)
+{
+    bool should_have_handle = true;
+
+    auto const surface = weak_surface.lock();
+    if (surface)
+    {
+        switch(surface->state())
+        {
+        case mir_window_state_attached:
+        case mir_window_state_hidden:
+            should_have_handle = false;
```

Comment:
This does raise the interesting question: should windows on other workspaces show up in the switcher, and how can that be accomodated? (I *think* the answer is “with a protocol revision”?)

Reply:
Seeing as the protocol makes no mention of workspaces, I think they should show up for now. If you wanted to filter based on workspaces then that would indeed require updates to the protocol.

Reply:
The way that miral-shell and egmde implement workspaces is to hide the windows of inactive workspaces. (But the Mir implementation of "workspaces" is just keeping lists of windows, the policy is what gives those lists meaning.)